ls1intum/Artemis
# `Development`: Add endpoints for build agent information in local continuous integration

### Checklist
#### General


#### Server

### Motivation and Context


For admins, it might be interesting to get information on available build agents in local CI, e.g. number of build jobs currently running on an agent, build job capacity.

### Description

This PR adds some REST endpoints to be later used for displaying build agent information on the client. Each hazelcast member adds itself to the list when the LocalCISharedBuildJobQueueService is initialized. The information is updated everytime a member processes a build job. When calling the endpoint, the list is updated to make sure that ex members who left the cluster will not be represented.

### Steps for Testing

Prerequisites:
- 1 Instructor
- Programming exercise
- Postman, logged in as admin user

1. In Postman, GET /api/admin/build-job-queue/build-agents 
    - On TS6, TS11 and locally, you should see one entry, on TS3 three entries (you might have to wait for a moments, until all nodes have registered)
    - `numberOfCurrentBuildJobs` should be 0 for all build agents
2. Create some submissions for the programming exercise
3. Check the endpoint again. Some build agents should have `numberOfCurrentBuildJobs` > 0


### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
